app/auth.py

```diff
@@ -0,0 +1,106 @@
+import os
+from enum import Enum
+from functools import wraps
+
+import requests
+from flask import Blueprint, current_app, flash, redirect, request, url_for
+from flask_login import UserMixin, current_user, login_user
+from flask_login.login_manager import LoginManager
+from flask_login.mixins import AnonymousUserMixin
+from oauthlib.oauth2 import WebApplicationClient
+
+authorization_base_url = "https://github.com/login/oauth/authorize"
+token_url = "https://github.com/login/oauth/access_token"
+user_url = "https://api.github.com/user"
+
+client = WebApplicationClient(os.environ.get("OAUTH_CLIENT_ID"))
+if os.environ.get("FLASK_ENV") == "development" or True:  # Heroku has no ssl
+    os.environ["OAUTHLIB_INSECURE_TRANSPORT"] = "1"
+
+
+class Role(Enum):
+    READER = "READER"
+    WRITER = "WRITER"
+
+
+roles = {
+    "JoelWee": Role.WRITER.value,
+}
+
+
+def init_config_manager(app):
+    login_manager = LoginManager()
+
+    if app.config.get("LOGIN_DISABLED"):
+        AnonymousUser.role = Role.WRITER.value
+    login_manager.anonymous_user = AnonymousUser
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        return redirect(url_for("auth.login"))
+
+    @login_manager.user_loader
+    def load_user(user_id):
+        r = requests.get(f"https://api.github.com/users/{user_id}")
+        user_data = r.json()
+        return User(user_data["login"])
```

Comment:
The `user_id` passed through here should already be the user's id/github username - so we shouldn't need to make a new request to Github on every user request. I.e. the below should suffice here!
```

    @login_manager.user_loader
    def load_user(user_id):
        return User(user_id)
```